app/services/duplicate_benchmark_resolver.rb

```diff
@@ -4,54 +4,89 @@
 class DuplicateBenchmarkResolver
   class << self
     def run!
-      each_benchmark do |bm|
-        if benchmarks[[bm.ref_id, bm.version]]
-          migrate_benchmark(benchmarks[[bm.ref_id, bm.version]], bm)
+      @benchmarks = nil
+      duplicate_benchmarks.find_each do |bm|
+        if existing_benchmark(bm)
+          migrate_benchmark(existing_benchmark(bm), bm)
         else
-          benchmarks[[bm.ref_id, bm.version]] = bm
+          self.existing_benchmark = bm
         end
       end
     end
 
     private
 
-    def migrate_benchmark(existing_bm, duplicate_bm)
-      logger.info(
-        "Duplicate benchmark found for #{existing_bm.id} - #{duplicate_bm.id}"
-      )
-      migrate_rules(existing_bm, duplicate_bm)
-      migrate_profiles(existing_bm, duplicate_bm)
-      duplicate_bm.delete
+    def existing_benchmark(benchmark)
+      benchmarks[[benchmark.ref_id, benchmark.version]]
     end
 
-    def each_benchmark
-      Xccdf::Benchmark.includes(:rules, :profiles).find_each do |bm|
-        Xccdf::Benchmark.transaction do
-          yield bm
-        end
-      end
+    def existing_benchmark=(benchmark)
+      benchmarks[[benchmark.ref_id, benchmark.version]] = benchmark
     end
 
     def benchmarks
       @benchmarks ||= {}
     end
 
+    def duplicate_benchmarks
+      Xccdf::Benchmark.joins(
+        "JOIN (#{grouped_nonunique_benchmark_tuples.to_sql}) as bm on "\
+        'benchmarks.ref_id = bm.ref_id AND '\
+        'benchmarks.version = bm.version'
+      )
+    end
+
+    def grouped_nonunique_benchmark_tuples
+      Xccdf::Benchmark.select(:ref_id, :version)
+                      .group(:ref_id, :version)
+                      .having('COUNT(id) > 1')
+    end
+
+    def migrate_benchmark(existing_bm, duplicate_bm)
+      logger.info(
+        "Duplicate benchmark found for #{existing_bm.id} - #{duplicate_bm.id}"
+      )
+      migrate_rules(existing_bm, duplicate_bm)
+      migrate_parent_profiles(existing_bm, duplicate_bm)
+      migrate_profiles(existing_bm, duplicate_bm)
+      remove_remaining_parent_profiles(duplicate_bm)
```

Comment:
Any profiles that are referenced as parent profiles by another profile cannot be deleted because of the foreign key constraint on the DB. This method clears out any other parent_profile_ids for those profiles we're about to delete.

Here's an example of the error from AR:

```
ActiveRecord::InvalidForeignKey: PG::ForeignKeyViolation: ERROR:  update or delete on table "profiles" violates foreign key constraint "fk_rails_ba3d1c4f86
" on table "profiles"                                                                                                                                                                         
        DETAIL:  Key (id)=(6021fe64-4c07-4f60-8edb-318ae06cf139) is still referenced from table "profiles".                                                                                   
        : DELETE FROM "profiles" WHERE "profiles"."id" = $1
```